README.md
# nock-back-ci

[![Build Status](https://travis-ci.org/joanvila/nock-back-ci.svg?branch=master)](https://travis-ci.org/joanvila/nock-back-ci)
[![Coverage Status](https://coveralls.io/repos/github/joanvila/nock-back-ci/badge.svg?branch=master)](https://coveralls.io/github/joanvila/nock-back-ci?branch=master)

A simple acceptance testing helper optimised for complex CI problems

## Motivation

As responsible developers we want to test our NodeJS services with acceptance tests to simulate real traffic.

In a continuous deployment environment, the CI pipeline should be able to run those acceptance tests, however, the deployment of our service shouldn’t depend on an external service being up and running. Moreover, to run our service, we may need to access other services that are not reachable from the CI pipeline.

## Solution

When the acceptance tests are run in local, they will query the real external services. During this run,
the responses of all the external http calls will be stored in a fixture file.

When the tests are run in the CI pipeline, instead of querying the external services,
we will use the mocked responses stored in the fixture file.

This process will be transparent for the developers maintaining the tests.

## Usage

For a complete working example check `examples/server.test.js`.

```javascript
const NockBackCI = require('nock-back-ci');
// appProvider is an awaitable function that returns your express app
const appProvider = require('./yourapp');

// The localEnvironment flag should be set to true when the environment is local
// and false in the CI environment. This is usually done via environment variables.
const nockBackCiConfig = {
  localEnvironment: true,
  fixtureName: 'exampleFixture.json',
  fixtureDir: path.join(__dirname, 'fixtures'),
};

const nockBackCI = new NockBackCI(nockBackCiConfig);
const server = await nockBackCI.bootServer(appProvider); // server is an instance of supertest request

const testCase = await nockBackCI.testCaseInit();

// Test your api here

nockBackCI.testCaseEnd(testCase);
nockBackCI.killServer(server, done);
```

**Tip**: For the best experience place the usage of `nock-back-ci` in the `before/after` jest functions.
Check the `examples/server.test.js` example for inspiration.

## Extra options

```javascript
const nockBackCiConfig = {
  localEnvironment: true,
  fixtureName: 'exampleFixture.json',
  fixtureDir: path.join(__dirname, 'fixtures'),
  whitelistedHosts: /(localhost|127\.0\.0\.1|kms.amazonaws)/,
  healthcheck: '/operations/healthcheck', // The test won't start until this endpoint replies a 200
  customFilter: (scope) => true,
};
```

### Custom filters

In order to avoid recording specific responses in the fixtures depending on their content, the `customFilter` extra option becomes handy.

For example, the following custom filter would avoid recording empty responses to the fixture file:

```javascript
const nockBackCiConfig = {
  customFilter: (scope) => scope.response,
}
```

### Security

Keep in mind that `nock-back-ci` will record ALL http calls made by your service by default.

Please take care if your application sends or receives sensitive information like credentials, access keys, or users' personal information so that these data are not committed to your repository inside the fixutes.

To skip nocking for a particular endpoint, add it with the `whitelistedHosts` option to the `nock-back-ci` config like the example above.

### API Warmup

In some cases, starting the api and having it ready to serve traffic may take a while.
This is usually the case when it queries external services and warms up caches.

If this is your case, by providing an optional `healthcheck` parameter the test won't start until the healthcheck of the api becomes green.

Moreover when using this functionality, a separate fixture named `boot.json` will be created to store
the http responses of the services queried at startup time.

## Todo

Items on the roadmap.

- Add tests to the lib/options file
- Feature: When to recreate local fixtures policy
- Feature: Config validation

## Contributing

See the `CONTRIBUTING.md` file.

## License

MIT License (c) 2018 Joan Vilà Cuñat


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 1.0.0 - 2019-02-12

### Changed
- Keeping the latest request response in the fixutres instead of the first
one in the case of duplicates

## 0.3.0 - 2019-02-12

### Added
- Custom filters to avoid recording requests depending on the response

## 0.2.0 - 2018-07-10

### Changed
- Waiting for the healthcheck to become green now is optional
- The healthcheck endpoint is configurable

## 0.1.1 - 2018-07-10

### Added
- Tests for the package
- CI pipeline with travis
- Coverage reports

## 0.1.0 - 2018-30-09

### Added
- First version of the package



test/index.test.js
const supertest = require('supertest');
const nock = require('nock');
const fs = require('fs');
const NockBackCI = require('../index');


jest.mock('nock', () => {
  const back = jest.fn();
  const restore = jest.fn();
  const enableNetConnect = jest.fn();
  return {
    back,
    restore,
    enableNetConnect,
  };
});

const mockServerGet = jest.fn();
const mockServer = {
  get: mockServerGet,
};

jest.mock('supertest', () => {
  const agent = jest.fn();
  return {
    agent,
  };
});

let existsSyncMock = null;
let unlinkSyncMock = null;
const nockDoneMock = jest.fn();

describe('NockBackCI', () => {
  beforeEach(() => {
    existsSyncMock = jest.spyOn(fs, 'existsSync');
    unlinkSyncMock = jest.spyOn(fs, 'unlinkSync');

    existsSyncMock.mockImplementation(() => true);
    unlinkSyncMock.mockImplementation(() => true);

    nock.back.mockImplementation(() => Promise.resolve({ nockDone: nockDoneMock }));
    nock.back.setMode = jest.fn();

    mockServerGet.mockImplementation(() => Promise.resolve({ statusCode: 200 }));
    supertest.agent.mockImplementation(() => mockServer);
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
  });

  const defaultConfig = {
    localEnvironment: true,
    fixtureName: 'fixture.json',
    fixtureDir: null,
    whitelistedHosts: /(localhost|127\.0\.0\.1)/,
    healthcheck: null,
    customFilter: () => true,
  };

  describe('constructor', () => {
    it('merges the default config with the user config', () => {
      const userConfig = {
        fixtureName: 'myFixture.json',
        fixtureDir: 'fixtureDir',
        healthcheck: '/operations/healthcheck',
        customFilter: () => false,
      };

      const nockBackCi = new NockBackCI(userConfig);

      expect(nockBackCi.config).toEqual({
        localEnvironment: defaultConfig.localEnvironment,
        fixtureName: 'myFixture.json',
        fixtureDir: 'fixtureDir',
        whitelistedHosts: defaultConfig.whitelistedHosts,
        healthcheck: '/operations/healthcheck',
        customFilter: expect.any(Function),
      });
      expect(nockBackCi.config.customFilter()).toBe(false);
    });

    it('sets nock back fixture directory', () => {
      const userConfig = {
        fixtureDir: 'fixtureDir',
      };

      new NockBackCI(userConfig); // eslint-disable-line no-new

      expect(nock.back.fixtures).toEqual('fixtureDir');
    });

    it('does not check if the file exists if we are not in development', () => {
      const userConfig = {
        localEnvironment: false,
      };

      new NockBackCI(userConfig); // eslint-disable-line no-new

      expect(existsSyncMock).not.toBeCalled();
    });

    it('removes the fixtures in local environment', () => {
      const userConfig = {
        localEnvironment: true,
        fixtureName: 'myFixture.json',
        fixtureDir: 'fixtureDir',
      };

      new NockBackCI(userConfig); // eslint-disable-line no-new

      expect(existsSyncMock).toBeCalledWith('fixtureDir/myFixture.json');
      expect(existsSyncMock).toBeCalledWith('fixtureDir/boot.json');
      expect(unlinkSyncMock).toBeCalledWith('fixtureDir/myFixture.json');
      expect(unlinkSyncMock).toBeCalledWith('fixtureDir/boot.json');
    });

    it('doesnt remove the fixtures if they do not exists', () => {
      existsSyncMock.mockImplementation(() => false);
      const userConfig = {
        localEnvironment: true,
        fixtureName: 'myFixture.json',
        fixtureDir: 'fixtureDir',
      };

      new NockBackCI(userConfig); // eslint-disable-line no-new

      expect(existsSyncMock).toBeCalledWith('fixtureDir/myFixture.json');
      expect(existsSyncMock).toBeCalledWith('fixtureDir/boot.json');
      expect(unlinkSyncMock).not.toBeCalled();
      expect(unlinkSyncMock).not.toBeCalled();
    });
  });

  describe('bootServer', () => {
    const userConfig = {
      fixtureName: 'myFixture.json',
      fixtureDir: 'fixtureDir',
    };

    const appProviderMock = jest.fn();

    beforeEach(() => {
      appProviderMock.mockImplementation(() => Promise.resolve('appMock'));
    });

    it('sets the nock back mode to record', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.bootServer(appProviderMock);

      expect(nock.back.setMode).toHaveBeenCalledWith('record');
    });

    it('starts recording fixtures', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.bootServer(appProviderMock);

      expect(nock.back).toHaveBeenCalledWith(
        'boot.json',
        expect.any(Object),
      );
    });

    it('enables the connection for the allowed hosts', async () => {
      const userConfigWithWhitelistedHosts = {
        ...userConfig,
        whitelistedHosts: defaultConfig.whitelistedHosts,
      };
      const nockBackCi = new NockBackCI(userConfigWithWhitelistedHosts);

      await nockBackCi.bootServer(appProviderMock);

      expect(nock.enableNetConnect).toHaveBeenCalledWith(
        defaultConfig.whitelistedHosts,
      );
    });

    it('returns the server created with the app provided by appProvider', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      const server = await nockBackCi.bootServer(appProviderMock);

      expect(appProviderMock).toHaveBeenCalledTimes(1);
      expect(supertest.agent).toHaveBeenCalledWith('appMock');
      expect(server).toBe(mockServer);
    });

    it('does not call the healthcheck to boot the app if the endpoint is not provided', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.bootServer(appProviderMock);

      expect(mockServerGet).not.toHaveBeenCalled();
    });

    it('calls the healthcheck to boot the app if the endpoint is provided', async () => {
      const configWithHealthcheck = {
        ...userConfig,
        healthcheck: '/healthcheck/endpoint',
      };
      const nockBackCi = new NockBackCI(configWithHealthcheck);

      await nockBackCi.bootServer(appProviderMock);

      expect(mockServerGet).toHaveBeenCalledWith('/healthcheck/endpoint');
    });

    it('calls nockDone', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.bootServer(appProviderMock);

      expect(nockDoneMock).toHaveBeenCalled();
    });

    it('sets the nock back mode to wild', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.bootServer(appProviderMock);

      expect(nock.back.setMode).toHaveBeenCalledWith('wild');
    });
  });

  describe('testCaseInit', async () => {
    const userConfig = {
      fixtureName: 'myFixture.json',
      fixtureDir: 'fixtureDir',
      whitelistedHosts: 'localhost',
    };

    it('sets the nock back mode to record', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.testCaseInit();

      expect(nock.back.setMode).toHaveBeenCalledWith('record');
    });

    it('starts recording fixtures', async () => {
      const nockBackCi = new NockBackCI(userConfig);

      await nockBackCi.testCaseInit();

      expect(nock.back).toHaveBeenCalledWith(
        userConfig.fixtureName,
        expect.any(Object),
      );
    });

    it('enables the connection for the allowed hosts', async () => {
      const userConfigWithWhitelistedHosts = {
        ...userConfig,
        whitelistedHosts: defaultConfig.whitelistedHosts,
      };
      const nockBackCi = new NockBackCI(userConfigWithWhitelistedHosts);

      await nockBackCi.testCaseInit();

      expect(nock.enableNetConnect).toHaveBeenCalledWith(
        defaultConfig.whitelistedHosts,
      );
    });

    it('returns the nockDone', async () => {
      const userConfigWithWhitelistedHosts = {
        ...userConfig,
        whitelistedHosts: defaultConfig.whitelistedHosts,
      };
      const nockBackCi = new NockBackCI(userConfigWithWhitelistedHosts);

      const nockDone = await nockBackCi.testCaseInit();

      expect(nockDone).toBe(nockDoneMock);
    });
  });

  describe('testCaseEnd', () => {
    const userConfig = {
      fixtureName: 'myFixture.json',
      fixtureDir: 'fixtureDir',
    };

    it('stops recording fixtures when called', () => {
      const nockDoneParam = jest.fn();
      const nockBackCi = new NockBackCI(userConfig);

      nockBackCi.testCaseEnd(nockDoneParam);

      expect(nockDoneParam).toHaveBeenCalled();
    });

    it('sets the nock back mode to wild', () => {
      const nockBackCi = new NockBackCI(userConfig);

      nockBackCi.testCaseEnd(jest.fn());

      expect(nock.back.setMode).toHaveBeenCalledWith('wild');
    });
  });

  describe('killServer', () => {
    const mockCloseApp = jest.fn();

    const userConfig = {
      fixtureName: 'myFixture.json',
      fixtureDir: 'fixtureDir',
    };

    const mockServerParam = {
      app: { close: mockCloseApp },
    };

    it('sets the nock back mode to wild', () => {
      const nockBackCi = new NockBackCI(userConfig);

      nockBackCi.killServer(mockServerParam, jest.fn());

      expect(nock.back.setMode).toHaveBeenCalledWith('wild');
    });

    it('closes the app', () => {
      const doneMock = jest.fn();
      const nockBackCi = new NockBackCI(userConfig);

      nockBackCi.killServer(mockServerParam, doneMock);

      expect(mockCloseApp).toHaveBeenCalledWith(doneMock);
    });

    it('restores nock setings', () => {
      const nockBackCi = new NockBackCI(userConfig);

      nockBackCi.killServer(mockServerParam, jest.fn());

      expect(nock.restore).toHaveBeenCalled();
    });
  });
});


examples/server.test.js
const path = require('path');
const NockBackCI = require('../index');
const appProvider = require('./server');

jest.unmock('nock');

const nockBackCiConfig = {
  localEnvironment: true, // This would depend on the environment
  fixtureName: 'exampleFixture.json',
  fixtureDir: path.join(__dirname, 'fixtures'),
  whitelistedHosts: /(localhost|127\.0\.0\.1|amazonaws)/,
  healthcheck: '/operations/healthcheck',
};

let server = null;
let testCase = null;
let nockBackCI = null;

describe('acceptance test', () => {
  beforeAll(async () => {
    nockBackCI = new NockBackCI(nockBackCiConfig);
    server = await nockBackCI.bootServer(appProvider);
  });

  beforeEach(async () => {
    testCase = await nockBackCI.testCaseInit();
  });

  afterEach(() => {
    nockBackCI.testCaseEnd(testCase);
  });

  afterAll((done) => {
    nockBackCI.killServer(server, done);
  });

  it('validates the contract for at least one hotel', async () => {
    const response = await server.get('/endpoint');
    expect(response.body.title).toEqual('London');
    expect(response.status).toBe(200);
  });
});


test/index.nockBack.test.js
jest.unmock('nock');
const fs = require('fs');
const nock = require('nock');
const NockBackCI = require('../index');

const existsSyncMock = jest.spyOn(fs, 'existsSync');
existsSyncMock.mockImplementation(() => false);

const nockBackImplementation = jest.fn();
nockBackImplementation.prototype.setMode = jest.fn();
// nockBackImplementation.setMode = jest.fn();

const nockBackMock = jest.spyOn(nock, 'back');
nockBackMock.mockImplementation(() => nockBackImplementation);

describe('NockBackCI', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.resetModules();
  });

  describe('bootServer - with nock.back()', () => {
    // TODO: Test nock.back()
  });

  describe('testCaseInit - with nock.back()', () => {
    const userConfig = {
      fixtureName: 'myFixture.json',
      fixtureDir: 'fixtureDir',
      whitelistedHosts: 'localhost',
    };

    it.skip('starts recording fixtures', () => {
      // FIXME: How to mock nock.back if it's an object
      // that can be called as a function???
      const nockBackCi = new NockBackCI(userConfig);

      nockBackCi.testCaseInit();

      expect(nockBackMock).toHaveBeenCalledWith(
        userConfig.fixtureName,
        expect.any(Object),
      );
    });

    // TODO: Test the rest of stuff
  });
});


lib/options.js
/* eslint-disable no-param-reassign */
const zlib = require('zlib');

const makeCompressedResponsesReadable = (scope) => {
  if (scope.rawHeaders.indexOf('gzip') > -1) {
    const fullResponseBody = scope.response
      && scope.response.reduce
      && scope.response.reduce((previous, current) => previous + current);
    try {
      scope.response = JSON.parse(zlib.gunzipSync(Buffer.from(fullResponseBody, 'hex')).toString('utf8'));
    } catch (e) {
      scope.response = '';
    }
  }
  delete scope.rawHeaders;
  return scope;
};

const compareByUrl = (scopeA, scopeB) => {
  const fullUrlA = `${scopeA.scope}${scopeA.path}`;
  const fullUrlB = `${scopeB.scope}${scopeB.path}`;

  if (fullUrlA > fullUrlB) {
    return 1;
  }
  if (fullUrlA < fullUrlB) {
    return -1;
  }
  return 0;
};

const removeDuplicates = (outputs) => {
  const obj = {};
  return Object.keys(outputs.reduce((prev, next) => {
    // if (!obj[next.path]) obj[next.path] = next; // This would keep the oldest value
    obj[next.path] = next; // This keeps the newest value
    return obj;
  }, obj)).map(i => obj[i]);
};

const is200 = scope => scope.status === 200;

module.exports = (whitelistedHosts, customFilter) => ({
  afterRecord: (outputs) => {
    const reducedOutput = removeDuplicates(outputs);
    return reducedOutput
      .filter(o => !o.scope.match(whitelistedHosts))
      .filter(is200)
      .map(makeCompressedResponsesReadable)
      .filter(customFilter)
      .sort(compareByUrl);
  },
  after: scope => scope.persist(),
});


index.js
/* eslint-disable class-methods-use-this */
const request = require('supertest');
const nock = require('nock');
const path = require('path');

const fs = require('fs');
const getDefaultOptions = require('./lib/options');

const defaultConfig = {
  localEnvironment: true,
  fixtureName: 'fixture.json',
  fixtureDir: null,
  whitelistedHosts: /(localhost|127\.0\.0\.1)/,
  healthcheck: null,
  customFilter: () => true,
};


class NockBackCI {
  constructor(userConfig) {
    this.config = { ...defaultConfig, ...userConfig };
    this.defaultOptions = getDefaultOptions(this.config.whitelistedHosts, this.config.customFilter);

    nock.back.fixtures = this.config.fixtureDir;

    if (this.config.localEnvironment) {
      const testFixtureFile = path.join(this.config.fixtureDir, this.config.fixtureName);
      if (fs.existsSync(testFixtureFile)) {
        fs.unlinkSync(testFixtureFile);
      }

      const bootFixtureFile = path.join(this.config.fixtureDir, 'boot.json');
      if (fs.existsSync(bootFixtureFile)) {
        fs.unlinkSync(bootFixtureFile);
      }
    }
  }

  async bootServer(appProvider) {
    let healthChecked = false;

    nock.back.setMode('record');
    const { nockDone } = await nock.back('boot.json', this.defaultOptions);
    nock.enableNetConnect(this.config.whitelistedHosts);

    const app = await appProvider();
    const server = request.agent(app);

    while (this.config.healthcheck && !healthChecked) {
      // eslint-disable-next-line no-await-in-loop
      const response = await server.get(this.config.healthcheck);
      if (response && response.statusCode === 200) {
        healthChecked = true;
      }
    }

    nockDone();
    nock.back.setMode('wild');
    return server;
  }

  async testCaseInit() {
    nock.back.setMode('record');
    const { nockDone } = await nock.back(this.config.fixtureName, this.defaultOptions);
    nock.enableNetConnect(this.config.whitelistedHosts);
    return nockDone;
  }

  testCaseEnd(nockDone) {
    nockDone();
    nock.back.setMode('wild');
  }

  killServer(server, done) {
    nock.back.setMode('wild');
    server.app.close(done);
    nock.restore();
  }
}

module.exports = NockBackCI;
